File: Week3/Python/ArrayProblems/NextPermutation.py
"""
    A permutation of an array of integers is an arrangement of its members into a sequence or linear order.
    The next permutation of an array of integers is the next lexicographically greater permutation of its integer.
    If such arrangement is not possible, the array must be rearranged as the lowest possible order (i.e., sorted in 
    ascending order).

    https://leetcode.com/problems/next-permutation/
"""

def nextPermutation(nums):
    """ T(c) -> O(n), S(c) -> O(1) """
    # using longest prefix match
    break_point = -1
    
    i = len(nums)-2
    while i >= 0:
        if nums[i] < nums[i+1]:
            break_point = i
            break
        i = i - 1
    
    # no break point found
    if break_point == -1:
        # reverse the array, as that permutation is last permutation possible, make it as first permutation
        nums.reverse()
    else:
        j = len(nums)-1
        # find smallest number which is bigger than break_point number
        while j > i:
            if nums[j] > nums[break_point]:
                # swap elements
                nums[j] , nums[break_point] = nums[break_point] , nums[j]
                break
            j = j - 1
        
        # reverse remaining elements after breakpoint
        start = break_point + 1
        end = len(nums)-1
        while start < end:
            nums[start], nums[end] = nums[end], nums[start]
            start += 1
            end -= 1


arr = [2, 1, 5, 4, 3, 0, 0]
print(arr)
nextPermutation(arr)
print(arr)
arr = [3, 2, 1]
print(arr)
nextPermutation(arr)
print(arr)

File: Week3/Java/ArrayProblem/NextPermutation.java
public class NextPermutation {

/*
 * A permutation of an array of integers is an arrangement of its members into a sequence or linear order.
 * The next permutation of an array of integers is the next lexicographically greater permutation of its integer.
 * If such arrangement is not possible, the array must be rearranged as the lowest possible order (i.e., sorted in ascending order).
 * 
 * https://leetcode.com/problems/next-permutation/
 * T(c) - > O(n), S(c) -> O(1)
 */
    static void nextPermutation(int[] nums) {
        int n = nums.length;
        int breakIdx = -1;
        
        // find breakpoint, first index i from the back of the given array where arr[i] < arr[i+1]
        for(int i=n-2;i>=0;i--)
        {
            if(nums[i]<nums[i+1])
            {
                breakIdx = i;
                break;
            }
        }

        // if there is no decrement/break point, reverse the whole array
        if(breakIdx == -1)
        {
            for(int i=0,j=n-1; i<j; i++,j--)
            {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
        }
        // if breakpoint found
        else{
            // find some number from right side which is small but greater than that number, swap
            for(int i=n-1;i>breakIdx;i--)
            {
                if(nums[i]>nums[breakIdx])
                {
                    // swap numbers
                    int temp = nums[i];
                    nums[i] = nums[breakIdx];
                    nums[breakIdx] = temp;
                    break;
                }
            }
            
            // sort/reverse remaining numbers (right side of that break point index)
            for(int i=breakIdx+1,j=n-1; i<j; i++,j--)
            {
                int temp = nums[i];
                nums[i] = nums[j];
                nums[j] = temp;
            }
        }
    }

    public static void main(String[] args) {
        int arr[] = {1,2,3};
        nextPermutation(arr);
        for(int i=0;i<arr.length;i++)
        {
            System.out.print(arr[i]+" ");
        }
        System.out.println();
    }
}
